phabricatoremails/render/template.py

```diff
@@ -0,0 +1,268 @@
+    if diff_type == DiffLineType.ADDED:
+        return "added-line"
+    elif diff_type == DiffLineType.REMOVED:
+        return "removed-line"
+    elif diff_type == DiffLineType.NO_CHANGE:
+        return "no-change-line"
+
+
+def reviewer_status_icon(status: ReviewerStatus):
+    if status == ReviewerStatus.ACCEPTED:
+        return emoji_html("check_mark")
+    elif status == ReviewerStatus.REQUESTED_CHANGES:
+        return emoji_html("deny_x")
+    else:
+        return ""
+
+
+def existence_change(reviewer_change: ExistenceChange):
+    if reviewer_change == ExistenceChange.NO_CHANGE:
+        return ""
+    elif reviewer_change == ExistenceChange.ADDED:
+        return "added"
+    elif reviewer_change == ExistenceChange.REMOVED:
+        return "removed"
+
+
+def file_change(change: AffectedFileChange):
+    if change == AffectedFileChange.MODIFIED:
+        return "modified"
+    elif change == AffectedFileChange.ADDED:
+        return "added"
+    elif change == AffectedFileChange.REMOVED:
+        return "removed"
+
+
+def diff_symbol(diff_type: DiffLineType):
+    if diff_type == DiffLineType.ADDED:
+        return "+"
+    elif diff_type == DiffLineType.REMOVED:
+        return "-"
+    elif diff_type == DiffLineType.NO_CHANGE:
+        return ""
+
+
+def text_comment(comment: str):
+    # Email width is 80, use two characters for "> ". So, each line gets 78 characters
+    return "\n".join(["> " + line for line in textwrap.wrap(comment, 78)])
+
+
+def text_reviewer_status(status: ReviewerStatus):
+    if status == ReviewerStatus.ACCEPTED:
+        return "(r+)"
+    elif status == ReviewerStatus.REQUESTED_CHANGES:
+        return "(r-)"
+    else:
+        return ""
+
+
+def text_existence_change(change: ExistenceChange):
+    if change == ExistenceChange.ADDED:
+        return "(added)"
+    elif change == ExistenceChange.REMOVED:
+        return "(removed)"
+    else:
+        return ""
+
+
+class Template:
+    """Renders the raw HTML and text Jinja templates
+
+
+    Additionally, for the rendered HTML, the CSS is inlined to improve compatibility
+    with complex email clients.
+    """
+
+    def __init__(self, html_template: jinja2.Template, text_template: jinja2.Template):
+        self.html_template = html_template
+        self.text_template = text_template
+
+    def render(self, css_inline: Premailer, template_params: Dict):
+        html = self.html_template.render(**template_params)
+        text = self.text_template.render(**template_params)
+
+        return css_inline.transform(html, False), text
+
+
+class TemplateStore:
+    """Configures Jinja and exposes the templates
+
+    There's two sets of templates: text and html. Each set has its own
+    distinct configuration and set of custom tests and filters.
+    """
+
+    def __init__(self):
+        self.html_jinja_env = self._jinja_html()
+        self.text_jinja_env = self._jinja_text()
+
+    @staticmethod
+    def _jinja_html():
```

Comment:
I've moved `_jinja_html()` and `_jinja_text()` outside of the class, that was a good recommendation :+1:.

I'm having issues splitting `template.py` apart. The template helper functions (like `is_reply`, `is_reviewer_soft_needed`, `emoji_html`, etc) are only useful for the jinja templates. I've thrown `_` in front of each of their names to make their "private-only usefulness" more clear.

I could move the `jinja_env` setup functions out too, but it needs access to these private template helper methods, and they're only used in a private-ish way by `TemplateStore`.

I think that if we were to modularize this section, it would probably be better to shard based on the domain - modularize the reviewer-specific helper functions, modularize the file-specific helper functions, etc. However, I don't think there's enough information to do this modularization intelligently yet, we'll have to see how this module evolves.
I don't think they're hurting anything by being collected all together just yet.